### PR TITLE
Discriminator/bill

### PR DESCRIPTION
**Pull request by billyang98**
Sorry, had some automatic formatting at the beginning. Ignore the first commit if it bothers you.

Use the if statements for backwards compatibility, since we need to expose more losses and the encoder hidden states when doing discriminator loss.

### CODE REVIEW

File: src/train.py [hunk 1]
```diff
@@ -97,36 +97,100 @@ def speech_loss(gold_mel, stop_label, pred_mel, stop_pred):
     stop_loss = F.binary_cross_entropy_with_logits(stop_pred.squeeze(), stop_label)
     return pred_loss + stop_loss
 
+
+def cross_entropy(input, target, size_average=True):
```
Comment (Lucaskabela) [inline src/train.py:101]: Why not just use the torch cross entropy here? Is it the smoothing term?  Why is the smoothing term important here? (I didn't read the original paper that closely...)

File: src/train.py [hunk 2]
```diff
@@ -97,36 +97,100 @@ def speech_loss(gold_mel, stop_label, pred_mel, stop_pred):
...
     x, y = process_batch(batch)
     character, mel, mel_input, _  = x
     gold_char, gold_mel, gold_stop = y
 
-    text_pred = model.text_ae(character).permute(1, 2, 0)
-    pred, stop_pred = model.speech_ae(mel, mel_input)
+    if use_dis_loss:
+        text_pred, t_hid = model.text_ae(character, use_dis_loss)
+        text_pred = text_pred.permute(1, 2, 0)
+        t_d_loss = discriminator_hidden_to_loss(model, t_hid, 'speech')
```
Comment (Lucaskabela) [inline src/train.py:161]: Wait, this is the text autoencoder... shouldn't it be 'text'?

File: src/train.py [hunk 3]
```diff
@@ -97,36 +97,100 @@ def speech_loss(gold_mel, stop_label, pred_mel, stop_pred):
...
+
+        pred, stop_pred, s_hid = model.speech_ae(mel, mel_input, use_dis_loss)
+        s_d_loss = discriminator_hidden_to_loss(model, s_hid, 'text')
```
Comment (Lucaskabela) [inline src/train.py:164]: Similar here - why are we using 'text' as the target for the seech autoencoder?

File: src/train.py [hunk 4]
```diff
@@ -97,36 +97,100 @@ def speech_loss(gold_mel, stop_label, pred_mel, stop_pred):
...
+def supervised_step(model, batch, use_dis_loss=False):
     x, y = process_batch(batch)
     character, mel, mel_input, _  = x
     gold_char, gold_mel, gold_stop = y
 
-    pred, stop_pred = model.tts(character, mel_input)
-    text_pred = model.asr(character, mel).permute(1, 2, 0)
+    if use_dis_loss:
+        pred, stop_pred, t_hid = model.tts(character, mel_input, ret_enc_hid=use_dis_loss)
+        t_d_loss = discriminator_hidden_to_loss(model, t_hid, 'speech')
```
Comment (Lucaskabela) [inline src/train.py:183]: Same issue here - if we are using "text" input, shouldn't the label be text?

File: src/train.py [hunk 5]
```diff
@@ -157,47 +241,106 @@ def optimizer_step(loss, model, optimizer, args):
...
 
     # Log losses
     losses['t_ae'].append(t_ae_loss.detach().cpu().item())
     losses['s_ae'].append(s_ae_loss.detach().cpu().item())
+    if args.use_discriminator:
+        losses['t_ae_d'].append(t_d_loss.detach().cpu().item())
+        losses['s_ae_d'].append(s_d_loss.detach().cpu().item())
 
 def train_cm_step(losses, model, optimizer, batch, args):
+    model.text_m.train()
```
Comment (Lucaskabela) [inline src/train.py]: For efficency sake, can we just set these once at the top of train?

**Review by Lucaskabela — CHANGES_REQUESTED** [review]

File: src/train.py (hunk 1 above)
Comment (billyang98) [reply src/train.py:101]: Yes, torch loss does not allow for soft targets. The smoothing term has been shown to give better convergence on classification problems and was used in the Almost Unsupervised paper.

File: src/train.py (hunk 2 above)
Comment (billyang98) [reply src/train.py:161]: The encoders should try to fool the discriminator, thus target should be 'speech'.

File: src/train.py (hunk 5 above)
Comment (billyang98) [reply src/train.py]: It cannot be set once at the top of train because we alternate training steps between the model and the discriminator. It can be set inside of the training loop, but the current code is more consistent in case we want to do some different training experiments without having to refactor the training loop again.

Also I don't believe this affects efficiency much, and this is how it is done in the FB MT code.

File: src/train.py (hunk 2 above)
Comment (Lucaskabela) [reply src/train.py:161]: Ah, got it since this is adversarial right

**Review by Lucaskabela — APPROVED** [review]
